File: src/main/java/com/turleylabs/algo/trader/kata/framework/SimpleMovingAverage.java
package com.turleylabs.algo.trader.kata.framework;

import java.util.LinkedList;

public class SimpleMovingAverage {
    LinkedList<Double> buffer = new LinkedList<>();
    private String symbol;
    private int length;

    public SimpleMovingAverage(String symbol, int length) {
        this.symbol = symbol;
        this.length = length;
    }

    public double getValue() {
        return this.buffer.stream().mapToDouble(Double::doubleValue).average().getAsDouble();
    }

    public boolean isReady() {
        return buffer.size() == length;
    }

    public void addData(double data) {
        this.buffer.add(data);

        if (this.buffer.size() > length)
        {
            this.buffer.removeFirst();
        }
    }

    public String getSymbol() {
        return symbol;
    }
}

File: typescript/src/framework/simple_moving_average.ts
export class SimpleMovingAverage {
  private buffer: number[] = [];
  constructor(readonly symbol: string, private length: number) {
  }

  get isReady(): boolean {
    return this.buffer.length === this.length;
  }

  get value(): number {
    const sum = this.buffer.reduce((a, b) => a + b, 0);
    const numElements = this.buffer.length;
    if (numElements === 0) {
      throw new Error('Average of empty list is undefined');
    }
    return sum / numElements;
  }

  addData(data: number) {
    this.buffer.push(data);
    if (this.buffer.length > this.length) {
      this.buffer.splice(0, 1);
    }
  }
}